calliope/backend/pyomo/constraints/group.py

```diff
@@ -1,892 +1,1416 @@
+        return return_noconstraint("carrier_prod", group_name)
+    else:
+        lhs_loc_tech_carriers = get_group_lhs_loc_tech_carriers(
+            backend_model, group_name
+        )
+
+        lhs = sum(
+            backend_model.carrier_prod[loc_tech_carrier, timestep]
+            for loc_tech_carrier in lhs_loc_tech_carriers
+            for timestep in backend_model.timesteps
+        )
+        return equalizer(lhs, limit, what)
+
+
+def target_reserve_share_constraint_rule(
+    backend_model, group_name, timestep, what
+):  # UPDATED the math
+    """
+        Enforces carrier_prod for groups of technologies and locations,
+        as a sum over the entire model period.
+
+        .. container:: scrolling-wrapper
+
+            .. math::
+
+    <<<<<<< HEAD
+                \\sum_{loc::tech \\in given\\_group, timestep \\in timesteps} cap\\_value(loc::tech, timestep) energy_{cap}(loc::tech) timestep\\_resolution(timestep) \\leq \\geq \\sum_{loc::tech::carrier \\in loc\\_tech\\_carriers_{in}, timestep \\in timesteps} (1 + target\\_share) carrier_{con}(loc::tech::carrier, timestep)
+    =======
+                \\sum_{loc::tech::carrier \\in given\\_group, timestep \\in timesteps} carrier_{prod}(loc::tech::carrier, timestep) \\leq carrier_prod_max
+    >>>>>>> parent of d7bd4705 (Update math for reserve margin constraints)
+
+    """
+
+    target_share = get_param(
+        backend_model, f"group_target_reserve_share_{what}", (group_name)
+    )
+
+    if invalid(target_share):
+        return return_noconstraint("target_reserve_share", group_name)
+    else:
+        lhs_loc_tech_carriers = get_group_lhs_and_rhs_loc_tech_carriers(
+            backend_model, group_name
+        )[0]
+
+        carrier = lhs_loc_tech_carriers[-1].split("::")[-1]
+        locs = [
+            loc_tech_carrier.split("::")[0]
+            for loc_tech_carrier in lhs_loc_tech_carriers
+        ]
+        timestep_resolution = backend_model.timestep_resolution[timestep]
+
+        lhs = sum(
+            (
+                get_param(
+                    backend_model,
+                    "cap_value",
+                    ((loc_tech_carrier.rsplit("::", 1)[0]), timestep),
+                )
+            )
+            * backend_model.energy_cap[loc_tech_carrier.rsplit("::", 1)[0]]
+            * timestep_resolution
+            for loc_tech_carrier in lhs_loc_tech_carriers
+        )
+
+        if hasattr(backend_model, "loc_techs_transmission"):
+            loc_tech_carriers_transmission = [
+                loc_tech_transmission + "::" + carrier
+                for loc_tech_transmission in backend_model.loc_techs_transmission
+            ]
+        else:
+            loc_tech_carriers_transmission = []
+
+        loc_tech_carriers_con_all = [
+            loc_tech_carrier
+            for loc_tech_carrier in backend_model.loc_tech_carriers_con
+            if loc_tech_carrier not in loc_tech_carriers_transmission
+        ]
+
+        rhs = (

```

Comment:
I've been running into issues with the reserve margins when transmission is included in the model. It looks like (both directions) of transmission is included in the energy_cap calculations on the lhs of the reserve constraints, so adding a reserve doesn't result in generation capacity that would meet the reserve requirements. If I set `target_reserve_share_equals` in the sample model below (a single supply at one location and demand at another) the model either doesn't solve or only builds the supply to 1/3 of the reserve requirement with` ensure_feasibility`. Not sure if it would be possible to leave transmission out by default but allow you to specify transmission techs in the constraint and have them be added in if desired.

If I try and filter out the transmission techs, the group constraint doesn't seem to work at all. Setting `target_resrve_share_equals` again with a technology filter (and enabling `ensure_feasibility`) results in the model building no supply at all even with unmet demand.

Looking into the code it looks like the issue is because the rhs of the share and adder constraints uses the `locs` list derived from `lhs_loc_tech_carriers` which doesn't contain the demand location if there is nothing in the lhs there. With transmission included the demand loc is included because of the transmission link, but filtering out transmission to try and get the model to build enough generation for the reserve removes that link and results in the loc not being included.

[sample_model.zip](https://github.com/calliope-project/calliope/files/14544028/sample_model.zip)

Reply:
The `group_cosntraint_loc_tech_carriers_{}_rhs` attribute also seems to not include demand loc_techs for the appropriate reserve constraints as the code only looks for "share" and "demand"/"import" in the constraint name (`constraint_sets.py:524`). Grabbing the rhs loc_techs from `get_group_lhs_and_rhs_loc_tech_carriers` would need that code adjusted so that demand and consumption techs are included in the rhs for the non-absolute reserve constraints.